File: android/src/main/java/expo/modules/signonglass/ExpoSignOnGlassModule.kt
package expo.modules.signonglass

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.functions.Queues

class ExpoSignOnGlassModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoSignOnGlass')` in JavaScript.
    Name("ExpoSignOnGlass")

    // Enables the module to be used as a native view. Definition components that are accepted as part of
    // the view definition: Prop, Events.
    View(ExpoSignOnGlassView::class) {
      AsyncFunction("expose") { view: ExpoSignOnGlassView, promise: Promise ->
        try {
          val image = view.getSignatureImage()
          promise.resolve(image)
        } catch (e: Exception) {
         promise.reject(CodedException(message = e.message ?: "Unknown error"))
        }
      }.runOnQueue(Queues.MAIN)

      AsyncFunction("clear") { view: ExpoSignOnGlassView, promise: Promise ->
        view.clearCanvas()
      }.runOnQueue(Queues.MAIN)

      Events("onStartSign")
    }
  }
}

File: ios/ExpoSignOnGlassModule.swift
import ExpoModulesCore

public class ExpoSignOnGlassModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    Name("ExpoSignOnGlass")

    View(ExpoSignOnGlassView.self) {
        AsyncFunction("expose") {  (view: ExpoSignOnGlassView, promise: Promise)  in
          if let image = view.getSignatureImage() {
          // Compress the image data with medium quality (0.5)
            if let compressedData = image.jpegData(compressionQuality: 0.5) {
              let base64String = compressedData.base64EncodedString()
              let dataUri = "data:image/jpeg;base64," + base64String
              promise.resolve(dataUri)
            } else {
              promise.reject(Exception(name: "CompressionError", description: "Failed to compress image"))
            }
          } else {
            promise.reject(Exception(name: "GetSignatureError", description: "Failed to get signature image"))
          }
        }.runOnQueue(.main)

        AsyncFunction("clear") { (view: ExpoSignOnGlassView, promise: Promise)  in
          view.clearCanvas()
        }.runOnQueue(.main)

        Events("onStartSign") 

        Prop("thickness") { (view: ExpoSignOnGlassView, thickness: CGFloat) in
          view.setThickness(thickness)
        }
    }

  }
}
